ipatests/test_integration/test_ipa_ipa_migration.py

```diff
@@ -946,3 +889,429 @@ def test_ipa_hbac_rule_duplication(self):
+        MIGRATION_COMPLETE_LOG_MSG = "Migration complete!\n"
+        result = run_migrate(
+            self.replicas[0],
+            "stage-mode",
+            self.master.hostname,
+            "cn=Directory Manager",
+            self.master.config.admin_password,
+            extra_args=['-n'],
+        )
+        install_msg = self.replicas[0].get_file_contents(
+            paths.IPA_MIGRATE_LOG, encoding="utf-8"
+        )
+        assert result.returncode == 0
+        assert MIGRATION_SCHEMA_LOG_MSG in install_msg
+        assert MIGRATION_CONFIG_LOG_MSG in install_msg
+        assert IPA_UPGRADE_LOG_MSG in install_msg
+        assert SIDGEN_TASK_LOG_MSG in install_msg
+        assert MIGRATION_COMPLETE_LOG_MSG in install_msg
+
+    def test_ipa_migrate_stage_mode_new_user(self):
+        """
+        This testcase checks that when a new user is added and
+        ipa-migrate is run in stage-mode, uid/gid of the
+        migrated user is not preserved i.e we have different
+        uid/gid for user on remote and local IPA server.
+        """
+        username = 'stageuser1'
+        original_passwd = 'Secret123'
+        base_dn = str(self.master.domain.basedn)
+        LOG_MSG1 = (
+            "DEBUG Resetting the DNA range for new entry: "
+            "uid={},cn=users,cn=accounts,{}\n"
+        ).format(username, base_dn)
+        tasks.user_add(self.master, username, password=original_passwd)
+        run_migrate(
+            self.replicas[0],
+            "stage-mode",
+            self.master.hostname,
+            "cn=Directory Manager",
+            self.master.config.admin_password,
+            extra_args=['-n']
+        )
+        install_msg = self.replicas[0].get_file_contents(
+            paths.IPA_MIGRATE_LOG, encoding="utf-8"
+        )
+        assert LOG_MSG1 not in install_msg
+        tasks.clear_sssd_cache(self.master)
+        self.master.run_command(['ipa', 'user-show', username])
+        cmd1 = self.master.run_command(['id', username])
+        tasks.clear_sssd_cache(self.replicas[0])
+        self.replicas[0].run_command(['ipa', 'user-show', username])
+        cmd2 = self.replicas[0].run_command(['id', username])
+        assert cmd1.stdout_text != cmd2.stdout_text
+
+
+class TestIPAMigrationProdMode(IntegrationTest):
+    """
+    Tests for ipa-migrate tool in prod mode
+    """
+    topology = "line"
+    num_replicas = 1
+    num_clients = 1
+
+    @classmethod
+    def install(cls, mh):
+        tasks.install_master(cls.master, setup_dns=True, setup_kra=True)
+        prepare_ipa_server(cls.master)
+        tasks.install_client(cls.master, cls.clients[0], nameservers=None)
+
+    def test_remote_server(self):
+        """
+        This test installs IPA server instead of replica on
+        system under test with the same realm and domain name.
+        """
+        tasks.install_master(self.replicas[0], setup_dns=True, setup_kra=True)
+
+    def test_ipa_migrate_prod_mode(self, empty_log_file):
```

Comment:
All the tests in this class call ipa-migrate with exactly the same options. To speed up things, you can call ipa-migrate only once and only check the outputs in the other tests:
```
    def test_ipa_migrate_prod_mode(self, empty_log_file):
        tasks.kinit_admin(self.master)
        tasks.kinit_admin(self.replicas[0])
        result = run_migrate(
            self.replicas[0],
            "prod-mode",
            self.master.hostname,
            "cn=Directory Manager",
            self.master.config.admin_password,
            extra_args=['-n'],
        )
        assert result.returncode == 0

    def test_prod_mode_logs(self):
        MIGRATION_SCHEMA_LOG_MSG = "Migrating schema ...\n"
        MIGRATION_DATABASE_LOG_MSG = (
            "Migrating database ... (this may take a while)\n"
        )
        IPA_UPGRADE_LOG_MSG = (
            "Running ipa-server-upgrade ... (this may take a while)\n"
        )
        SIDGEN_TASK_LOG_MSG = "Running SIDGEN task ...\n"
        install_msg = self.replicas[0].get_file_contents(
            paths.IPA_MIGRATE_LOG, encoding="utf-8"
        )
        assert MIGRATION_SCHEMA_LOG_MSG in install_msg
        assert MIGRATION_DATABASE_LOG_MSG in install_msg
        assert IPA_UPGRADE_LOG_MSG in install_msg
        assert SIDGEN_TASK_LOG_MSG in install_msg

    def test_ipa_migrate_prod_mode_habc_rule(self):
        hbac_rule_name1 = 'test1'
        hbac_rule_name2 = 'testuser_sshd'

        cmd1 = self.replicas[0].run_command(
            ["ipa", "hbacrule-find", hbac_rule_name1])
        cmd2 = self.replicas[0].run_command(
            ["ipa", "hbacrule-find", hbac_rule_name2])
        assert hbac_rule_name1 in cmd1.stdout_text
        assert hbac_rule_name2 in cmd2.stdout_text
 ...
```

Do the same for the class in Stage mode.